source/common/stats/symbol_table_impl.h

```diff
@@ -170,30 +187,53 @@ class SymbolTable {
    *
    * @param symbol_vec the vector of symbols to be freed.
    */
-  void free(StatName stat_name);
+  void free(const StatName& stat_name);
+
+  /**
+   * StatName backing-store can be managed by callers in a variety of ways
+   * to minimize overhead. But any persistent reference to a StatName needs
+   * to hold onto its own reference-counts for all symbols. This method
+   * helps callers ensure the symbol-storage is maintained for the lifetime
+   * of a reference.
+   *
+   * @param symbol_vec the vector of symbols to be freed.
+   */
+  void incRefCount(const StatName& stat_name);
+
+#ifndef ENVOY_CONFIG_COVERAGE
+  // It is convenient when debugging to be able to print the state of the table,
+  // but this code is not hit during tests ordinarily, and is not needed in
+  // production code.
+  void debugPrint() const;
+#endif
+
+  /**
+   * Decodes a vector of symbols back into its period-delimited stat name. If
+   * decoding fails on any part of the symbol_vec, we release_assert, since this
+   * should never happen, and we don't want to continue running with a corrupt
+   * stats set.
+   *
+   * @param symbol_vec the vector of symbols to decode.
+   * @return std::string the retrieved stat name.
+   */
+  std::string decode(const SymbolStorage symbol_vec, uint64_t size) const;
 
 private:
   friend class StatName;
   friend class StatNameTest;
 
   struct SharedSymbol {
+    SharedSymbol(Symbol symbol) : symbol_(symbol), ref_count_(1) {}
+
     Symbol symbol_;
-    uint32_t ref_count_;
+    std::atomic<uint32_t> ref_count_;
   };
 
   // This must be called during both encode() and free().
-  mutable Thread::MutexBasicLockable lock_;
+  // mutable Thread::MutexBasicLockable lock_;]
```

Comment:
nit: should this be uncommented or removed?

Reply:
removed